src/lcm/LifeCycleStates.cc

```diff
@@ -2496,3 +2681,504 @@ void LifeCycleManager::trigger_resize_failure(int vid)
+        /* ------------------------------------------------------------------ */
+        ds_deltas.replace("SIZE", reserved_sz - real_sz);
+        ds_deltas.replace("IMAGES", 0);
+
+        Quotas::ds_del(vm_uid, vm_gid, &ds_deltas);
+
+        /* ------------------------------------------------------------------ */
+        /* Remove VM from Backup Job pending list                             */
+        /* ------------------------------------------------------------------ */
+        if (auto bj = bjpool->get(bj_id))
+        {
+            bj->backup_finished(vid, true);
+
+            bjpool->update(bj.get());
+        }
+
+        return;
+
+error_increment_update:
+        if (incremental_id == -1)
+        {
+            backups.incremental_backup_id(-1);
+            backups.del(image_id);
+        }
+
+error_image_create:
+        vm->log("LCM", Log::ERROR, "backup_success, " + error);
+        vmpool->update(vm.get());
+
+        vm.reset();
+
+        Quotas::ds_del(vm_uid, vm_gid, &ds_deltas);
+        return;
+    });
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+void LifeCycleManager::trigger_backup_failure(int vid)
+{
+    trigger([this, vid]
+    {
+        int vm_uid{0}, vm_gid{0}, bj_id{-1};
+        Template ds_deltas;
+
+        if ( auto vm = vmpool->get(vid) )
+        {
+            switch(vm->get_lcm_state())
+            {
+                case VirtualMachine::BACKUP:
+                    vm->set_state(VirtualMachine::RUNNING);
+                    break;
+
+                case VirtualMachine::BACKUP_POWEROFF:
+                    vm->set_state(VirtualMachine::POWEROFF);
+                    vm->set_state(VirtualMachine::LCM_INIT);
+                    break;
+
+                default:
+                    vm->log("LCM", Log::ERROR, "backup_failure, VM in a wrong state");
+                    break;
+            }
+
+            vm_uid = vm->get_uid();
+            vm_gid = vm->get_gid();
+
+            auto& backups = vm->backups();
+            int incremental_id = backups.incremental_backup_id();
+            Backups::Mode mode = backups.mode();
+            bj_id = backups.backup_job_id();
+
+            vm->backup_size(ds_deltas);
+            ds_deltas.add("DATASTORE", vm->backups().last_datastore_id());
+
+            if (mode == Backups::FULL || incremental_id == -1)
+            {
+                ds_deltas.add("IMAGES", 1);
+            }
+
+            backups.last_backup_clear();
+
+            vmpool->update(vm.get());
+        }
+
+        // Quota rollback
+        Quotas::ds_del(vm_uid, vm_gid, &ds_deltas);
+
+        /* Remove VM from Backup Job pending list                             */
+        /* ------------------------------------------------------------------ */
+        if (auto bj = bjpool->get(bj_id))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

We observed that you are using unintended operations or assignments that could lead to logical errors, vulnerabilities, or undefined behavior. To prevent these issues, ensure that you follow best practices, such as using explicit comparisons and clear operator usage, and using parentheses to clarify intent. Learn more - https://cvefeed.io/cwe/detail/cwe-480-use-of-incorrect-operator